Sources/SwiftProtobuf/Google_Protobuf_Value+Extensions.swift

```diff
@@ -76,6 +76,7 @@ extension Google_Protobuf_Value: _CustomJSONCodable {
   }
 
   internal mutating func decodeJSON(from decoder: inout JSONDecoder) throws {
```

Comment:
The interactions at this level are the part that give me the most pause and get hard to test.

So a Value that is:
- just a boolean/string/number == depth of 1
- an object, becomes a Struct, so by the time you get the the values within the object, they are now 3?  (outer Value, Struct, and then inner Value(s))
- a list, becomes a List, so again, by the time you get to the values in the list, they are also at 3.

Since Struct/List contain objects, I feel like they really should factor into depth, but I have a harder time with the rest.  Yes, from a pure object model pov, they are all objects, but from a recursion budget, I'm not completely sure what to do.

thoughts?

Reply:
The recursion budget is necessarily an approximation.  We really care about stack space, but:
* We can't (portably) know how much stack space is available.
* We can't measure stack space used, just recursive calls.

So I would not count recursions for simple values (since they don't really recurse).  I would just count a recursion for every `[` or `{` parsed.  The drawback of course is that it may require some extra thought to come up with good tests:  Maybe we could have a test that appends `[` and `{` randomly, and repeats the decode after each append.  If our recursion tracking is working correctly, that should never crash.

Reply:
If we count "nesting" instead of "recursion", I think this can be a lot simpler.   Then we basically just increment for each `[` and `{` and decrement for `]` and `}`.  We should not be recursing for any other JSON elements, so that should provide the same net effect (and be simpler to explain and simpler to test).

Reply:
ack.  Going to resolve this since I think the discussion on the `*ArrayObject*` apis capture the issue as we close in to the decision.